File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-0.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        unordered_map <Node*, Node* > m; 
        m[NULL] = NULL;


        Node* cur = head; 
        while(cur){
            Node* copy = new Node(cur->val);
            m[cur] = copy;
            cur = cur->next;
        }
        
        cur = head; 
        while(cur){
            Node* copy = m[cur];
            copy->next = m[cur->next];
            copy->random = m[cur->random];
            cur = cur->next;
        }


        return m[head]; 
        
    }
};

File: Data Structures & Algorithms/clone-graph/submission-0.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        create = {}
        if not node:
            return node
        create[node] = Node(node.val)

        q = deque()
        q.append(node)

        while q:
            cur = q.popleft()
            for i in cur.neighbors:
                if i not in create:
                    create[i] = Node(i.val)
                    q.append(i)
                create[cur].neighbors.append(create[i])

        return create[node]

    
            


            

        